src/quoting.rs

```diff
@@ -0,0 +1,185 @@
+
+impl ToTokens for Program {
+  fn to_tokens(&self, tokens: &mut TokenStream) {
+    let constants = &self.values;
+    let functions = &self.functions;
+    let cases = functions.iter().map(|function| {
+      let name = format_ident!("F_{}", function.name);
+
+      quote! { #name => self.#name(book, ptr, argument), }
+    });
+
+    tokens.append_all(quote! {
+      use crate::run::*;
+
+      #( #constants )*
+
+      impl Net {
+        #( #functions )*
+
+        pub fn call_native(&mut self, book: &Book, ptr: Ptr, argument: Ptr) -> bool {
+          match ptr.val() {
+            #( #cases )*
+            _ => { return false; }
+          }
+        }
+      }
+    })
+  }
+}
+
+impl ToTokens for Function {
+  fn to_tokens(&self, tokens: &mut TokenStream) {
+    let name = format_ident!("F_{}", self.name);
+    let body = &self.body;
+
+    tokens.append_all(quote! {
+      pub fn #name(&mut self, book: &Book, ptr: Ptr, argument: Ptr) -> bool {
+        #( #body )*
+      }
+    })
+  }
+}
+
+impl ToTokens for TypeRepr {
+  fn to_tokens(&self, tokens: &mut TokenStream) {
+    tokens.append_all(match self {
+      TypeRepr::HvmPtr => quote! { Ptr },
+      TypeRepr::Ptr => quote! { usize },
+      TypeRepr::USize => quote! { usize },
+      TypeRepr::U8 => quote! { u8 },
+      TypeRepr::U32 => quote! { u32 },
+      TypeRepr::Bool => quote! { bool },
+    })
+  }
+}
+
+impl ToTokens for Stmt {
+  fn to_tokens(&self, tokens: &mut TokenStream) {
+    tokens.append_all(match self {
+      Stmt::Let { name, value } => {
+        let name = format_ident!("{name}");
+        quote! { let #name = #value; }
+      },
+      Stmt::Val { name, type_repr } => {
+        let name = format_ident!("{name}");
+        quote! { let #name: #type_repr; }
+      },
+      Stmt::Assign { name, value } => match name {
+        Prop::Anni => quote! { self.anni = #value; },
+        Prop::Oper => quote! { self.oper = #value; },
+        Prop::Eras => quote! { self.eras = #value; },
+        Prop::Comm => quote! { self.comm = #value; },
+        Prop::Var(var) => {
+          let name = format_ident!("{}", var).to_token_stream();
+          quote! { #name = #value; }
+        },
+      },
+      Stmt::Instr(instr) => quote! { #instr; },
+      Stmt::Free(value) => quote! { self.free(#value); },
+      Stmt::Return(value) => quote! { return #value; },
+      Stmt::SetHeap { idx, port, value } => quote! { self.heap.set(#idx, #port, #value); },
+      Stmt::Link { lhs, rhs } => quote! { self.link(#lhs, #rhs); },
+    })
+  }
+}
+
+impl ToTokens for Instr {
+  fn to_tokens(&self, tokens: &mut TokenStream) {
+    tokens.append_all(match self {
+      Instr::True => quote! { true },
+      Instr::False => quote! { true },
+      Instr::Int(i) => TokenStream::from_str(&format!("{i}")).unwrap(),
+      Instr::Const(Const::F(name)) => format_ident!("F_{}", name).into_token_stream(),
+      Instr::Const(Const::P1) => quote! { crate::run::P1 },
```

Comment:
Why not have `Const` hold just the number since that's what it will compile to?
And then separate the function constant into a different `Instr` variant

Reply:
easier to translate to different targets